File: api_server/platform_info.py
import psutil
import platform
import subprocess
import re
import functools
from middleware import transform_to


def want_str(s):
    if isinstance(s, bytes):
        return s.decode()
    return str(s)


def run_once(func):  # Result will not be affected by arguments.
    result = None

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        nonlocal result
        if result is not None:
            return result

        result = func(*args, **kwargs)
        return result

    return wrapper


class RoseliaSysInfo:
    @classmethod
    def cpu_count(cls, logical=True):
        return psutil.cpu_count(logical)

    @classmethod
    @run_once
    @transform_to(want_str)
    def processor_name(cls):
        if psutil.WINDOWS:
            family = platform.processor()
            name = want_str(subprocess.check_output(["wmic", "cpu", "get", "name"]).strip()).split("\n")[1]
            return ' '.join([name, family])
        elif psutil.MACOS:
            command = "/usr/sbin/sysctl -n machdep.cpu.brand_string"
            return subprocess.check_output(command.split(' ')).strip()
        elif psutil.LINUX:
            command = "cat /proc/cpuinfo"
            all_info = want_str(subprocess.check_output(command, shell=True).strip())
            for line in all_info.split("\n"):
                if "model name" in line:
                    return re.sub(r".*model name.*:", "", line, 1)
            else:
                return all_info
        return ""

    @classmethod
    @run_once
    def processor_info(cls):
        return {
            'name': cls.processor_name(),
            'total_core': cls.cpu_count(False),
            'logical_core': cls.cpu_count(True)
        }

    @classmethod
    @run_once
    def os_full_name(cls):
        constants = platform.uname()
        os_name = constants.system
        os_version = constants.version
        if psutil.MACOS:
            mac_version = platform.mac_ver()[0]
            os_version = mac_version
            v0, *v1 = map(int, mac_version.split('.'))
            v1 = v1[0]
            if v0 == 10:
                if v1 < 8:
                    os_name = 'MAC OS X'
                elif v1 < 12:
                    os_name = 'OS X'
                else:
                    os_name = 'macOS'

        return {
            'name': os_name,
            'version': os_version,
            'platform': platform.platform(1, 1),
            'node': constants.node,
            'server_interpreter': platform.python_implementation() + ' ' + platform.python_version()
        }

    @classmethod
    def memory_usage(cls):
        usage = psutil.virtual_memory()
        return {
            'total': usage.total,
            'used': usage.used,
            'percent': usage.percent,
            'available': usage.available
        }

    @classmethod
    def cpu_usage(cls):
        return {
            'total': psutil.cpu_percent(),
            'logic': psutil.cpu_percent(percpu=True)
        }

File: RoseliaCore/PlatformInfo.fs
module RoseliaBlog.RoseliaCore.PlatformInfo

open System
open System.Diagnostics
open System.Text.RegularExpressions

type PlatformInfo = {
    Name: string
    Version: string
    Node: string
    ServerInterpreter: string
} with
    static member internal GetCurrentMachine =
        let name =
            if OperatingSystem.IsWindows() then
                "Windows"
            else if OperatingSystem.IsMacOS() then
                match Environment.OSVersion.Version.Major with
                | v when v > 10 -> "macOS"
                | 10 ->
                    match Environment.OSVersion.Version.Minor with
                    | x when x < 8 -> "MAC OS X"
                    | x when x < 12 -> "OS X"
                    | _ -> "macOS"
                | _ -> "Mac OS"
            else if OperatingSystem.IsLinux() then
                "Linus"
            else "Unknown"
        
        {
            Name = name
            Version = Environment.OSVersion.VersionString
            Node = Environment.MachineName
            ServerInterpreter = Runtime.InteropServices.RuntimeInformation.FrameworkDescription
        }


type CpuInfo = {
    Name: string
    TotalCore: int
    LogicalCore: int
}

let private CpuCounter =
    lazy new PerformanceCounter("Processor", "% Processor Time", "_Total", true)
type CpuUsage = {
    Total: double
    Logic: double array
} with
    static member GetUsage =
        let mutable value = CpuCounter.Value.NextValue() |> float
        if value <= 0.0 then value <- CpuCounter.Value.NextValue() |> float
        {
            Total = value
            Logic = [||]
        }
        
        
type MemoryUsage = {
    Total: int64
    Used: int64
    Percent: double
    Available: int64
} with
    static member GetUsage =
        let proc = Process.GetCurrentProcess ()
        let memInfo = GC.GetGCMemoryInfo ()
        let used = GC.GetTotalMemory(false)
        let total = proc.PrivateMemorySize64
        
        {
            Available = memInfo.TotalAvailableMemoryBytes
            Used = used
            Percent = double(used / total)
            Total = total
        }

let CurrentPlatformInfo = lazy PlatformInfo.GetCurrentMachine

let private ExecuteAndGetOutput (command : string seq) =
    use p = new Process()
    p.StartInfo.RedirectStandardInput <- true
    p.StartInfo.RedirectStandardOutput <- true
    p.StartInfo.CreateNoWindow <- true
    p.StartInfo.FileName <- command |> Seq.head
    p.StartInfo.Arguments <-
        command
        |> Seq.tail
        |> String.concat " "
    
    p.Start() |> ignore
    let outcome = p.StandardOutput.ReadToEnd()
    p.WaitForExit()
    
    outcome

let private CpuName =
    lazy
        if OperatingSystem.IsWindows() then
            let output = ExecuteAndGetOutput ["wmic"; "cpu"; "get"; "name"]
            output.Trim().Split('\n').[1]
        else if OperatingSystem.IsMacOS() then
            let output = ExecuteAndGetOutput ["/usr/sbin/sysctl"; "-n"; "machdep.cpu.brand_string"]
            output.Trim()
        else if OperatingSystem.IsLinux() then
            let output = ExecuteAndGetOutput ["cat"; "/proc/cpuinfo"]
            seq {
                for line in output.Split('\n') do
                    if line.Contains "model name" then
                        yield Regex(".*model name.*:").Replace(line, "", 1)
            }
            |> Seq.tryHead
            |> Option.defaultValue ""
        else ""
        
let CurrentCpuInfo =
    lazy
        {
            Name = CpuName.Value
            // There is not solution for getting physical core count.
            // So I assume all processors support hyper threading.
            TotalCore = Environment.ProcessorCount <<< 1
            LogicalCore = Environment.ProcessorCount
        }
